mypy/dmypy.py

```diff
@@ -0,0 +1,585 @@
+# @action(subparse) decorator.
+
+parser = argparse.ArgumentParser(description="Client for mymy daemon mode",
+                                 fromfile_prefix_chars='@')
+parser.set_defaults(action=None)
+subparsers = parser.add_subparsers()
+
+start_parser = subparsers.add_parser('start', help="Start daemon")
+start_parser.add_argument('flags', metavar='FLAG', nargs='*', type=str,
+                          help="Regular mypy flags (precede with --)")
+
+status_parser = subparsers.add_parser('status', help="Show daemon status")
+
+stop_parser = subparsers.add_parser('stop', help="Stop daemon (asks it politely to go away)")
+
+kill_parser = subparsers.add_parser('kill', help="Kill daemon (kills the process)")
+
+restart_parser = subparsers.add_parser('restart',
+    help="Restart daemon (stop or kill followed by start)")
+restart_parser.add_argument('flags', metavar='FLAG', nargs='*', type=str,
+                            help="Regular mypy flags (precede with --)")
+
+check_parser = subparsers.add_parser('check', help="Check some files (requires running daemon)")
+check_parser.add_argument('-q', '--quiet', action='store_true',
+                          help="Suppress instrumentation stats")
+check_parser.add_argument('files', metavar='FILE', nargs='+', help="File (or directory) to check")
+
+recheck_parser = subparsers.add_parser('recheck',
+    help="Check the same files as the most previous  check run (requires running daemon)")
+recheck_parser.add_argument('-q', '--quiet', action='store_true',
+                            help="Suppress instrumentation stats")
+
+hang_parser = subparsers.add_parser('hang', help="Hang for 100 seconds")
+
+daemon_parser = subparsers.add_parser('daemon', help="Run daemon in foreground")
+daemon_parser.add_argument('flags', metavar='FLAG', nargs='*', type=str,
+                           help="Regular mypy flags (precede with --)")
+
+help_parser = subparsers.add_parser('help')
+
+
+def main() -> None:
+    """The code is top-down."""
+    args = parser.parse_args()
+    if not args.action:
+        parser.print_usage()
+    else:
+        args.action(args)
+
+
+ActionFunction = Callable[[argparse.Namespace], None]
+
+
+def action(subparser: argparse.ArgumentParser) -> Callable[[ActionFunction], None]:
+    """Decorator to tie an action function to a subparser."""
+    def register(func: ActionFunction) -> None:
+        subparser.set_defaults(action=func)
+    return register
+
+
+# Action functions (run in client from command line).
+
+@action(start_parser)
+def do_start(args: argparse.Namespace) -> None:
+    """Start daemon (it must not already be running).
+
+    This is where mypy flags are set.  Setting flags is a bit awkward;
+    you have to use e.g.:
+
+      dmypy start -- --strict
+
+    since we don't want to duplicate mypy's huge list of flags.
+    """
+    try:
+        pid, sockname = get_status()
+    except SystemExit as err:
```

Comment:
I'd prefer having a separate exception class for daemon errors, and a top-level try/except that will translate it to `SystemExit`.